internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -320,159 +277,59 @@ func (bt *bucketTicker) add(bucketID string) {
-							nb.parent.removeBucket(nbKey)
-							delete(bt.nodes, nbKey)
-							// At this point, add() should no longer be callable from the parent,
-							// so mark this nodeBatcher as dead (for adds that have already been called and are racing)
-							// and try to cut a final batch (just in case something was added in between
-							// the above if and marking this nodeBatcher as dead)
-							atomic.StoreUint32(&nb.dead, nodeStatusDead)
-							// Reload batch in case `add` loaded a new batch that was then added to.
-							currBatchPtr := (*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))
-							b = (*batch)(atomic.LoadPointer(currBatchPtr))
-							if b.getCurrentItemCount() > 0 {
-								nb.cutBatch(b)
-							}
-						}
-					}
-				}
-			case newBucketKey := <-bt.pendingNodes:
-				bt.nodes[newBucketKey] = true
-			case <-bt.stopCn:
-				bt.ticker.Stop()
-				return
-			}
-		}
-	})
+	bt.once.Do(bt.runTicker)
 }
 
-func (bt *bucketTicker) stop() {
-	close(bt.stopCn)
-}
-
-type batch struct {
-	items         atomic.Value // []*tracepb.Span
-	currCap       uint32
-	nextEmptyItem uint32
-	sendItemsSize uint32
-
-	closed  uint32
-	growMu  sync.Mutex
-	pending int32
-}
-
-func newBatch(initCapacity uint32, sendBatchSize uint32) *batch {
-	batch := &batch{
-		currCap:       initCapacity,
-		nextEmptyItem: uint32(0),
-		sendItemsSize: sendBatchSize,
-
-		closed: uint32(0),
-	}
-
-	batch.items.Store(make([]*tracepb.Span, initCapacity))
-	return batch
-}
-
-func (b *batch) add(spans []*tracepb.Span) (cut bool, closed bool) {
-	if atomic.LoadUint32(&b.closed) == batchStatusClosed {
-		return false, true
-	}
-	// Add to pending and then check closed again. If closed, just return and decrement pending.
-	atomic.AddInt32(&b.pending, int32(1))
-	defer atomic.AddInt32(&b.pending, int32(-1))
-	if atomic.LoadUint32(&b.closed) == batchStatusClosed {
-		return false, true
-	}
-	openTill := atomic.AddUint32(&b.nextEmptyItem, uint32(len(spans)))
-	openFrom := openTill - uint32(len(spans))
-	currCap := atomic.LoadUint32(&b.currCap)
-	if openTill > currCap {
-		b.grow(openTill)
-	}
-
-	for spanIndex, span := range spans {
-		b.items.Load().([]*tracepb.Span)[openFrom+uint32(spanIndex)] = span
-	}
-
-	return openTill > b.sendItemsSize, false
-}
-
-func (b *batch) closeBatch() {
-	atomic.StoreUint32(&b.closed, batchStatusClosed)
+func (bt *bucketTicker) runTicker() {
 	for {
-		// We only need to wait for goroutines currently executing `add`
-		// We are safe from closing during a items mutation due to the second
-		// closed check after incrementing pending
-		pending := atomic.LoadInt32(&b.pending)
-		if pending == int32(0) {
+		select {
+		case <-bt.ticker.C:
+			for nbKey := range bt.nodes {
+				nb := bt.parent.getBucket(nbKey)
+				// Need to check nil here incase the node was deleted from the parent batcher, but
+				// not deleted from bt.nodes
+				if nb == nil {
+					// re-delete just in case
+					delete(bt.nodes, nbKey)
+					continue
```

Comment:
nit: since you have continue you don't need an `else`

Reply:
removed continue, since i think having the else is more clear